### DISCUSSION TITLE
How do I generate a new Code-Block from a function inside an R-Package?

### DISCUSSION THREAD

**Discussion by OffByOne37** [quarto-dev/quarto-cli#7887] [Q&A] [answered]
### Description

Hello there!

I'm currently seeking a one-line method to include TIKZ pictures from an external file within my .qmd file. 
I show you what I`m currently using:

`// Headers etc. 

'''{r}
tikz_path <- my_package::create_qmd(file_path = "../tex_code.tex")
'''

'''{r, child=tikz_path}

'''`

Here's what I'm doing :

- Using a function from my package, I generate a .qmd file from a .tex file that contains the necessary lines of code.

- Manually, I use the 'child' attribute in a new code chunk to include this generated .qmd file and display the TIKZ picture.

However, I aim to automate the insertion and execution of this 'child' code chunk to avoid manual intervention. Is there a way to programmatically generate and execute this 'child' code chunk in the .qmd file using R or any other method?

Any insights, suggestions, or alternative approaches would be greatly appreciated!

Thanks in advance for your help!

(Note: As it's my package, I'm open to modifying the code and return type of the function if necessary.)

**Comment by cderv** [comment]
In case you don't know, there is a `tikz` engine in **knitr** that is supported in Quarto too
https://github.com/yihui/knitr-examples/blob/master/058-engine-tikz.Rmd

<details>
<summary>Qmd example </summary>


````markdown
---
title: Fix asis_output
format: pdf
keep-md: true
engine: knitr
---

# TikZ graphics

## Description

The engine inserts the code into a latex-string-template, which is then processed by LaTeX (and the **magick** package if `fig.ext` is not `pdf`).

## Options

You can pass some options to the engine by defining `engine.opts`, e.g. use your own template instead of the default one to include the tikz code: `engine.opts = list(template = "mytemplate.tex")`. The default template can be found under `system.file('misc', 'tikz2pdf.tex', package = 'knitr')`.

## Example

An example of the tikz-engine from <https://raw.github.com/sdiehl/cats/master/misc/example.md>

```{tikz, tikz-ex, fig.cap = "Funky tikz", fig.ext = 'png', cache=TRUE}
\usetikzlibrary{arrows}
\begin{tikzpicture}[node distance=2cm, auto,>=latex', thick, scale = 0.5]
\node (P) {$P$};
\node (B) [right of=P] {$B$};
\node (A) [below of=P] {$A$};
\node (C) [below of=B] {$C$};
\node (P1) [node distance=1.4cm, left of=P, above of=P] {$\hat{P}$};
\draw[->] (P) to node {$f$} (B);
\draw[->] (P) to node [swap] {$g$} (A);
\draw[->] (A) to node [swap] {$f$} (C);
\draw[->] (B) to node {$g$} (C);
\draw[->, bend right] (P1) to node [swap] {$\hat{g}$} (A);
\draw[->, bend left] (P1) to node {$\hat{f}$} (B);
\draw[->, dashed] (P1) to node {$k$} (P);
\end{tikzpicture}
```

## Tips

To develop the tikz-code, you could use `qtikz` or `ktikz`.
````

</details>

This would be one way to insert a tikz plot. 

> Using a function from my package, I generate a .qmd file from a .tex file that contains the necessary lines of code.

Why do you need to convert tex into .qmd ? If you already have a `.tex` content, you can include as raw content into your document. 

````{=latex}
<content of your tex file>
````

Using `{{< include file >}}` would be a way to include it in the main .qmd (https://quarto.org/docs/authoring/includes.html)

> However, I aim to automate the insertion and execution of this 'child' code chunk to avoid manual intervention. Is there a way to programmatically generate and execute this 'child' code chunk in the .qmd file using R or any other method?

What would you like to type in your document ? 

You could probably leverage R to create a single function that would create the right markdown markup to insert  asis in the document. 

The cookbook for R Markdown shows how: https://bookdown.org/yihui/rmarkdown-cookbook/results-asis.html 
Quarto equivalent would be `output: asis`: https://quarto.org/docs/computations/execution-options.html#raw-output

You could create a function that create the right character vector to include in the document, and do 

````markdown

Some content

`r mypackage::add_tikz_graph()`

Some other content
````

And the inline code part would be a character vector of Raw markdown

You can also do it as a chunk 

````markdown
```{r}
#| echo: false
mypackage::add_tikz_graph()
```
````

Using `knitr::asis_output()` for the result of your function, and being sure to output the right markdown content. (`keep-md: true` in YAML header will allow you to check for the md content generated)

Hope it helps

**Reply by OffByOne37** [reply]
Yes i know that there exists a TIKZ-engine, and exactly that's the reason why I do not simply include the .tex-file. 

I know from the "include"-Option, but here i have the exact same problem, as with the "child"-code-chunk: i need two different code chunks/lines. 

I know from the "asis"-Option, but i was never able to run the generated code.
So my code now returns the Code with the corresponding code-chunk-header ({r, engine="tikz"}....) as Character-Vector. 

```
some content

`r mypackage::add_tikz_graph()`

some content

```

But now i simply have the whole code with 
```
{r, engine='tikz'}, #| label: fig-my-figure, #| echo: false, #| fig-align: center, #| fig-cap: Versuch TIKZ .......
```

on my output-slide, but not the generated tikz-picture? 

How exactly can i execute this Character-Vector?

**Reply by cderv** [reply]
It seems you have tried a lot already. Can you share a reproducible document or project ? 

You may just be missing new line here in the output you produce with your R function.

**Reply by cderv** [reply]
> So my code now returns the Code with the corresponding code-chunk-header ({r, engine="tikz"}....) as Character-Vector.µ

I also understand that your code contains the tikz engine now... so it indeed means you need to run as child 

Either using the `child = ` in chunk or just `knit_child()` https://bookdown.org/yihui/rmarkdown-cookbook/child-document.html

**Reply by cderv** [reply]
Example: 

````markdown
---
title: Fix asis_output
format: pdf
keep-md: true
engine: knitr
---

# TikZ graphics

```{r}
#| include: false
add_tikz_graph <- function() {
  res <- knitr::knit_child(text = glue::glue(r"(
             ```{tikz}
             #| label: fig-tikz-ex
             #| fig-cap: Funky tikz
             #| fig-ext: png
             \usetikzlibrary{arrows}
             \begin{tikzpicture}[node distance=2cm, auto,>=latex', thick, scale = 0.5]
             \node (P) {$P$};
             \node (B) [right of=P] {$B$};
             \node (A) [below of=P] {$A$};
             \node (C) [below of=B] {$C$};
             \node (P1) [node distance=1.4cm, left of=P, above of=P] {$\hat{P}$};
             \draw[->] (P) to node {$f$} (B);
             \draw[->] (P) to node [swap] {$g$} (A);
             \draw[->] (A) to node [swap] {$f$} (C);
             \draw[->] (B) to node {$g$} (C);
             \draw[->, bend right] (P1) to node [swap] {$\hat{g}$} (A);
             \draw[->, bend left] (P1) to node {$\hat{f}$} (B);
             \draw[->, dashed] (P1) to node {$k$} (P);
             \end{tikzpicture}
             ```)", .open = "<<", .close = ">>"), quiet = TRUE)
  knitr::asis_output(res)
}
```

`r add_tikz_graph()`

````

**Comment by OffByOne37** [comment, marked as answer]
Thank you!!!!

The provided example did not exactly what I wanted, since it included the asis_output in the slides, but i resolved the issue and could use this code to implement it in my package. Using the glue::glue function is also awesome, since then i can use an options-parameter to generate multiple options to the tikz-code!

Amazing support!

**Reply by cderv** [reply]
Great glad I could help !
